cpu/atxmega/periph/spi.c

```diff
@@ -0,0 +1,267 @@
+{
+    uint8_t *out_buf = (uint8_t *)out;
+    uint8_t *in_buf = (uint8_t *)in;
+
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        _print_buffer("bytes - out", out, len);
+    }
+
+    if (cs != SPI_CS_UNDEF) {
+        gpio_clear((gpio_t) cs);
+    }
+
+    for (size_t i = 0; i < len; i++) {
+        uint8_t tmp = (out_buf) ? out_buf[i] : 0;
+
+        dev(bus)->DATA = tmp;
+
+        while (!(dev(bus)->STATUS & SPI_IF_bm)){}
+
+        tmp = dev(bus)->DATA;
+
+        if (in_buf) {
+            in_buf[i] = tmp;
+        }
+    }
+
+    if ((!cont) && (cs != SPI_CS_UNDEF)) {
+        gpio_set((gpio_t) cs);
+    }
+
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        _print_buffer("bytes - in", in, len);
+    }
+}
+
+uint8_t spi_transfer_byte(spi_t bus, spi_cs_t cs, bool cont, uint8_t out)
+{
+    uint8_t in;
+
+    if (cs != SPI_CS_UNDEF) {
+        gpio_clear((gpio_t) cs);
+    }
+
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        _print_buffer("byte - out", &out, 1);
+    }
+
+    dev(bus)->DATA = out;
+
+    while (!(dev(bus)->STATUS & SPI_IF_bm)){}
+
+    if ((!cont) && (cs != SPI_CS_UNDEF)) {
+        gpio_set((gpio_t) cs);
+    }
+
+    in = dev(bus)->DATA;
+
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        _print_buffer("byte - in", &in, 1);
+    }
+
+    return in;
+}
+
+void spi_transfer_regs(spi_t bus, spi_cs_t cs, uint8_t reg,
+                       const void *out, void *in, size_t len)
+{
+    uint8_t *out_buf = (uint8_t *)out;
+    uint8_t *in_buf = (uint8_t *)in;
+
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        _print_buffer("regs - reg", &reg, 1);
+        _print_buffer("regs - out", out, len);
+    }
+
+    if (cs != SPI_CS_UNDEF) {
+        gpio_clear((gpio_t) cs);
+    }
+
+    dev(bus)->DATA = reg;
+
+    while (!(dev(bus)->STATUS & SPI_IF_bm)){}
+
+    for (size_t i = 0; i < len; i++) {
+        uint8_t tmp = (out_buf) ? out_buf[i] : 0;
+
+        dev(bus)->DATA = tmp;
+
+        while (!(dev(bus)->STATUS & SPI_IF_bm)){}
+
+        tmp = dev(bus)->DATA;
+
+        if (in_buf) {
+            in_buf[i] = tmp;
+        }
+    }
+
+    if (cs != SPI_CS_UNDEF) {
+        gpio_set((gpio_t) cs);
+    }
+
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        _print_buffer("regs - in", in, len);
+    }
+}
+
+uint8_t spi_transfer_reg(spi_t bus, spi_cs_t cs, uint8_t reg, uint8_t out)
+{
+    uint8_t in;
+
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        _print_buffer("reg - reg", &reg, 1);
+        _print_buffer("reg - out", &out, 1);
+    }
+
+    if (cs != SPI_CS_UNDEF) {
+        gpio_clear((gpio_t) cs);
+    }
+
+    dev(bus)->DATA = reg;
+
+    while (!(dev(bus)->STATUS & SPI_IF_bm)){}
+
+    (void)dev(bus)->DATA;   /* clear SPI_IF_bm flag */
+    dev(bus)->DATA = out;
+
+    while (!(dev(bus)->STATUS & SPI_IF_bm)){}
+
+    if (cs != SPI_CS_UNDEF) {
+        gpio_set((gpio_t) cs);
+    }
+
+    in = dev(bus)->DATA;
+
+    if (IS_ACTIVE(ENABLE_DEBUG)) {
+        _print_buffer("reg - in", &in, 1);
+    }
+
+    return in;
+}
```

Comment:
Note that we only provide these implementations, if the common implementations from `drivers/periph_common` cannot technically be used. Otherwise the common implementations typically use less ROM and in any case are easier to maintain.